src/main/java/com/mraof/minestuck/block/CassettePlayerBlock.java

```diff
@@ -0,0 +1,129 @@
+package com.mraof.minestuck.block;
+
+import com.mraof.minestuck.item.CassetteItem;
+import com.mraof.minestuck.tileentity.CassettePlayerTileEntity;
+import com.mraof.minestuck.util.CustomVoxelShape;
+import net.minecraft.block.Block;
+import net.minecraft.block.BlockState;
+import net.minecraft.entity.item.ItemEntity;
+import net.minecraft.entity.player.PlayerEntity;
+import net.minecraft.item.Item;
+import net.minecraft.item.ItemStack;
+import net.minecraft.state.BooleanProperty;
+import net.minecraft.state.StateContainer;
+import net.minecraft.tileentity.TileEntity;
+import net.minecraft.tileentity.TileEntityType;
+import net.minecraft.util.Hand;
+import net.minecraft.util.math.BlockPos;
+import net.minecraft.util.math.BlockRayTraceResult;
+import net.minecraft.world.IBlockReader;
+import net.minecraft.world.IWorld;
+import net.minecraft.world.World;
+
+import javax.annotation.Nullable;
+import java.util.function.Supplier;
+
+
+public class CassettePlayerBlock extends DecorBlock {
+    public static final BooleanProperty HAS_CASSETTE = ModBlockStateProperties.HAS_CASSETTE;
+    public static final BooleanProperty IS_OPEN = ModBlockStateProperties.IS_OPEN;
+    private final Supplier<TileEntityType<?>> tileType;
+
+    public CassettePlayerBlock(Properties properties, CustomVoxelShape shape, Supplier<TileEntityType<?>> tileType) {
+        super(properties, shape);
+        this.setDefaultState(this.stateContainer.getBaseState().with(HAS_CASSETTE, Boolean.valueOf(false)));
+        this.tileType = tileType;
+    }
+
+    @Override
+    @SuppressWarnings("deprecation")
+    public boolean onBlockActivated(BlockState state, World worldIn, BlockPos pos, PlayerEntity player, Hand handIn, BlockRayTraceResult hit) {
+        if (player.isSneaking()) {
+            if (state.get(IS_OPEN)) state = state.with(IS_OPEN, Boolean.valueOf(false));
+            else state = state.with(IS_OPEN, Boolean.valueOf(true));
+            worldIn.setBlockState(pos, state, 2);
+            return true;
+        } else if (state.get(HAS_CASSETTE) && state.get(IS_OPEN)) {
+            this.dropCassette(worldIn, pos);
+            state = state.with(HAS_CASSETTE, Boolean.valueOf(false));
+            worldIn.setBlockState(pos, state, 2);
+            return true;
+        } else {
+            return false;
+        }
+    }
+
+    public void insertCassette(IWorld worldIn, BlockPos pos, BlockState state, ItemStack cassetteStack) {
+        TileEntity tileentity = worldIn.getTileEntity(pos);
+        if (tileentity instanceof CassettePlayerTileEntity && state.get(IS_OPEN)) {
```

Comment:
Due to the "onBlockActivate" part, if the cassette player already have a cassette and the player right click, the cassette will get out

Reply:
But this function is called independently from `onBlockActivated()`, so that shouldn't matter for the implementation of `insertCassette()`

Reply:
I tried it in game, and yeah the cassette get out of the cassette player normally when you right click it with a cassette in hand

Reply:
Yeah, it still works because of other parts of the code (I think the item does that check before calling `insertCassette()`), but it still makes sense to have that check here alongside the `state.get(IS_OPEN)` for robustness reasons